api/proto/teleport/legacy/types/types.proto

```diff
@@ -1034,6 +1039,17 @@ message Header {
   string Value = 2 [(gogoproto.jsontag) = "value"];
 }
 
+// PortRange describes a port range for TCP apps. The range starts with Port and ends with EndPort.
+// PortRange can be used to describe a single port in which case the Port field is the port and the
+// EndPort field is 0.
+message PortRange {
+  // Port describes the start of the range. It must be between 1-65535.
+  uint32 Port = 1 [(gogoproto.jsontag) = "port"];
+  // EndPort describes the end of the range, inclusive. It must be between 2-65535 and be greater
+  // than Port when describing a port range. When describing a single port, it must be set to 0.
+  uint32 EndPort = 2 [(gogoproto.jsontag) = "end_port,omitempty"];
```

Comment:
maybe transforming this field into optional field?

I wonder if we shouldn't have
- Port
- StartPort
- EndPort

We validate so you can't specify Port and any of StartPort/Endport as it reads far easier or transform this into a oneof Port vs Range.

Reply:
> maybe transforming this field into optional field?

How exactly? My impression was that this field is already optional, as when describing a single port you can omit it. I should change the comment though to spell that out loud explicitly.

---

Regarding `StartPort`, I don't think `Port` & `EndPort` reads that terribly to warrant using one of the other two options. With three fields in a single struct, there's more invalid states that can be represented. With a oneof, as far as I understand, we'd lose some simplicity in `lib/config` & `lib/service/servicecfg` mirroring the structures defined in `api/types`. oneof in protos necessitates a wrapper struct which we'd have to replicate in those two packages.

Or I guess it could be something like this:

```
type PortSpec struct {
  Port Port
  PortRange PortRange
} 
```

…or some variation of it. But then again we'd have more stuff to validate and those three packages would no longer mirror their data structures in this specific case. Which I'm not sure how worthy of a goal it is to strive for, but it seems to me like a trait worth having.

Reply:
> How exactly? My impression was that this field is already optional, as when describing a single port you can omit it. I should change the comment though to spell that out loud explicitly.

Proto3 recently introduced the optional keyword that transforms the field into *fieldType. It's just a simpler way to say it out loud that it's not mandatory.

----

My idea is to keep it as simple as possible and be able to evolve independently. I don't think it reads bad, I just wonder if we want to evolve them in the future separately